src/Microsoft.OData.Core/ODataMetadataJsonOutputContext.cs

```diff
@@ -175,22 +175,53 @@ internal static Utf8JsonWriter CreateJsonWriter(Stream stream, ODataMessageWrite
         /// </summary>
         /// <param name="disposing">If 'true' this method is called from user code; if 'false' it is called by the runtime.</param>
         protected override void Dispose(bool disposing)
+        {
+            if (disposing)
+            {
+                try
+                {
+                    if (this.jsonWriter != null)
+                    {
+                        if (this.asynchronousOutputStream != null)
+                        {
+                            DisposeOutputStreamAsync().Wait();
+                        }
+                        else
+                        {
+                            this.jsonWriter.Flush();
+                            this.jsonWriter.Dispose();
+                        }
+
+                        this.messageOutputStream.Dispose();
+                    }
+                }
+                finally
+                {
+                    this.messageOutputStream = null;
+                    this.asynchronousOutputStream = null;
+                    this.jsonWriter = null;
+                }
+            }
+
+            base.Dispose(disposing);
+        }
+
+#if NETCOREAPP3_1_OR_GREATER
+        protected override async ValueTask DisposeAsyncCore()
         {
             try
             {
                 if (this.jsonWriter != null)
                 {
+                    await this.jsonWriter.FlushAsync().ConfigureAwait(false);

```

Comment:
The logic here is a bit different from the similar one in `Dispose` method. In `Dispose` method, these two statements are in an `else` block corresponding to `if (this.asynchronousOutputStream != null)`

Reply:
This `Dispose()` method first checks whether `this.asynchronousOutputStream` is not null, if it's not null, it calls `DisposeOutputStreamAsync()` method which asynchronously flushes and disposes the writer, then calls `Wait()` on the async method, otherwise it performs synchronously flush and dispose in the `else` block.. In `DisposeAsyncCore()` I didn't bother with synchronous flushes and disposal of the `JsonWriter`.